src/els_formatting_provider.erl

```diff
@@ -0,0 +1,112 @@
+
+%%==============================================================================
+%% els_provider functions
+%%==============================================================================
+
+
+%% Keep the behaviour happy
+-spec is_enabled() -> boolean().
+is_enabled() -> is_enabled_document().
+
+-spec is_enabled_document() -> boolean().
+is_enabled_document() -> true.
+
+-spec is_enabled_range() -> boolean().
+is_enabled_range() ->
+  false.
+
+%% NOTE: because erlang_ls does not send incremental document changes
+%%       via `textDocument/didChange`, this kind of formatting does not
+%%       make sense.
+-spec is_enabled_on_type() -> document_ontypeformatting_options().
+is_enabled_on_type() -> false.
+
+-spec handle_request(any(), els_provider:state()) ->
+  {any(), els_provider:state()}.
+handle_request({document_formatting, Params}, State) ->
+  lager:info("els_formatting_provide:document_formatting: [Params=~p]"
+            , [Params]),
+  #{ <<"options">>      := Options
+   , <<"textDocument">> := #{<<"uri">> := Uri}
+   } = Params,
+  lager:info("els_formatting_provide:document_formatting: [Options=~p]"
+            , [Options]),
+  {ok, Document} = els_utils:find_document(Uri),
+  case format_document(Uri, Document, Options) of
+    {ok, TextEdit} -> {TextEdit, State}
+  end;
+handle_request({document_rangeformatting, Params}, State) ->
+  #{ <<"range">>     := #{ <<"start">> := StartPos
+                         , <<"end">>   := EndPos
+                         }
+   , <<"options">>      := Options
+   , <<"textDocument">> := #{<<"uri">> := Uri}
+   } = Params,
+  Range = #{ start => StartPos, 'end' => EndPos },
+  {ok, Document} = els_utils:find_document(Uri),
+  case rangeformat_document(Uri, Document, Range, Options) of
+    {ok, TextEdit} -> {TextEdit, State}
+  end;
+handle_request({document_ontypeformatting, Params}, State) ->
+  #{ <<"position">>     := #{ <<"line">>      := Line
+                            , <<"character">> := Character
+                            }
+   , <<"ch">>           := Char
+   , <<"options">>      := Options
+   , <<"textDocument">> := #{<<"uri">> := Uri}
+   } = Params,
+  {ok, Document} = els_utils:find_document(Uri),
+  case ontypeformat_document(Uri, Document, Line + 1, Character + 1, Char
+                            , Options) of
+    {ok, TextEdit} -> {TextEdit, State}
+  end.
+
+%%==============================================================================
+%% Internal functions
+%%==============================================================================
+
+-spec format_document(uri(), els_document:document(), formatting_options())
+                     -> {ok, [text_edit()]}.
+format_document(Uri, _Document, Options) ->
+    lager:info("format_document: ~p", [{Uri, Options}]),
+    Path = els_uri:path(Uri),
+    Fun = fun(Dir) ->
+            RelPath = els_utils:project_relative(Uri),
+            OutFile = filename:join(Dir, RelPath),
+            Opts = #{output_dir => Dir},
```

Comment:
Shouldn't we pass to the formatter options such as `break_indent` and `sub_indent` calculated according to the client preferences?

Reply:
I guess we could. I saw this as initial support, to be used as a basis for the coming `erlfmt` from @michalmuskala.  So it is bare bones.

I need to rebase for the new persistent db, will see about adding it at the same time.

Reply:
Cool. In Emacs formatting has to be triggered via a specific command. What about VSCode? If the formatting is always requested explicitly by the client, I'm fine with merging this (after the tab/space options are passed to the formatter). If they are triggered by things like `TAB` or happen automatically in VSCode I would lean towards merging it but disabling it by default (until the formatter is good enough).